jme3-core/src/main/java/com/jme3/anim/tween/action/BlendAction.java

```diff
@@ -1,129 +1,265 @@
+     * Each action should have its own timeFactor used for interpolation.
+     * @param actions the actions, used for fetching the timeFactor values.
+     */
+    private void gatherTimeFactorPerAction(final BlendableAction... actions){
+        timeFactor = new double[actions.length];
+        for (int i = 0; i < actions.length; i++) {
+            //get the action duration.
+            final double actionLength = actions[i].getLength();
+            if (actionLength > 0 && getLength() > 0) {
+                //get the timeFactor(timeScale) of the inscribed action with respect to its parent Action.
+                this.timeFactor[i] = actionLength / getLength();
+            }else{
+                //assign a default value if durations are negative
+                this.timeFactor[i] = 1;
+            }
+        }
+    }
+
+    /**
+     * Used for gathering the targets & prepare them from the parsed anim actions args.
+     * @param actions the actions used for gathering targets.
+     */
+    private void gatherTargetsFromActions(final BlendableAction... actions){
+        for (final BlendableAction action : actions) {
+            //stretch the blend action length(duration) to the max possible length to accommodate the actions.
             if (action.getLength() > getLength()) {
                 setLength(action.getLength());
             }
-            Collection<HasLocalTransform> targets = action.getTargets();
-            for (HasLocalTransform target : targets) {
+            //get the target objects to apply the tracks extracted from the actions on them.
+            //NB: this line calls getTargets(); implementations on : ClipAction Class & Your Custom Actions(if exist) to fetch targets.
+            final Collection<HasLocalTransform> targets = action.getTargets();
+            //iterate over them
+            for (final HasLocalTransform target : targets) {
+                //get their current local transform
                 Transform t = targetMap.get(target);
+                //fill the empty map with empty transforms ready to collect interpolations from targets transforms.
                 if (t == null) {
                     t = new Transform();
                     targetMap.put(target, t);
                 }
             }
         }
-
-        //Blending effect maybe unexpected when blended animation don't have the same length
-        //Stretching any action that doesn't have the same length.
-        for (int i = 0; i < this.actions.length; i++) {
-            this.timeFactor[i] = 1;
-            if (this.actions[i].getLength() != getLength()) {
-                double actionLength = this.actions[i].getLength();
-                if (actionLength > 0 && getLength() > 0) {
-                    this.timeFactor[i] = this.actions[i].getLength() / getLength();
-                }
-            }
-        }
     }
 
+    /**
+     * The factory method, overridden to manage the interpolation.
```

Comment:
How is `doInterpolate()` a factory method?

Reply:
I haven't applied the changes requested by Paul on BlendAction yet.